File: apps/connections/native/android/app/src/main/java/com/fireblocks/connect/FireblocksConnect.kt
package com.fireblocks.connect

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.browser.customtabs.CustomTabsIntent
import java.security.SecureRandom

// ── Result ───────────────────────────────────────────────────────────────────

/** A wallet the user connected through the hosted Fireblocks flow. */
data class WalletConnection(
    val address: String,
    /** "evm" | "solana" */
    val chain: String,
    val walletName: String,
    /** Icon URL — usually an SVG-sprite URL, so render it in a WebView <img>. */
    val walletImage: String,
    /** The flow's session id, echoed back on the callback. Log it so a support
     *  report can be tied to a specific attempt. */
    val sessionId: String,
)

sealed class FireblocksConnectResult {
    data class Success(val wallet: WalletConnection) : FireblocksConnectResult()
    /** User dismissed the browser, or the flow reported a cancel. */
    object Cancelled : FireblocksConnectResult()
    /** Nonce mismatch (possible CSRF), an unparseable return, or a failure the
     *  flow reported. `code` is a stable machine code (e.g. "user_rejected")
     *  when the flow reported one; `sessionId` names the attempt. */
    data class Error(val reason: String, val code: String? = null, val sessionId: String? = null) :
        FireblocksConnectResult()
}

// ── Flow ─────────────────────────────────────────────────────────────────────

/**
 * Connect a self-custodial wallet through a hosted Fireblocks page — no SDK.
 *
 * ```kotlin
 * FireblocksConnect.present(
 *     context = this,
 *     flowURL = "https://connect.example.com/",
 *     scheme = "myapp",          // must match the intent-filter in AndroidManifest
 * ) { result ->
 *     when (result) {
 *         is FireblocksConnectResult.Success -> { /* result.wallet.address, .chain … */ }
 *         is FireblocksConnectResult.Cancelled -> {}
 *         is FireblocksConnectResult.Error -> { /* result.reason */ }
 *     }
 * }
 * ```
 *
 * Opens the page in a **Chrome Custom Tab** — Android's secure, sandboxed in-app
 * browser (the analog of iOS `ASWebAuthenticationSession`). The page returns to
 * `<scheme>://wallet-callback?address=…&nonce=…`, which `FireblocksRedirectActivity`
 * catches (see the AndroidManifest snippet in the README). This object appends
 * `redirect_uri` + a random `nonce` + `embedded=1`, verifies the nonce, and calls
 * your callback.
 *
 * Single in-flight connection at a time; the callback is held statically. For a
 * production app, prefer holding it in a ViewModel so it survives process death.
 */
object FireblocksConnect {
    private const val CALLBACK_HOST = "wallet-callback"

    private var pendingNonce: String? = null
    private var pendingCallback: ((FireblocksConnectResult) -> Unit)? = null

    fun present(
        context: Context,
        flowURL: String,
        scheme: String,
        onResult: (FireblocksConnectResult) -> Unit,
    ) {
        pendingNonce = randomNonce()
        pendingCallback = onResult
        val url = buildUrl(flowURL, scheme, pendingNonce!!)
        CustomTabsIntent.Builder().build().launchUrl(context, Uri.parse(url))
    }

    /** Called by [FireblocksRedirectActivity] when the return URL fires. */
    internal fun handleRedirect(uri: Uri) {
        val callback = pendingCallback ?: return
        val expected = pendingNonce
        pendingCallback = null
        pendingNonce = null

        if (uri.getQueryParameter("nonce") != expected) {
            callback(FireblocksConnectResult.Error("nonce mismatch"))
            return
        }
        // The flow reports failures through this same callback (status=error /
        // cancelled) so we learn the outcome and which session it was.
        val sessionId = uri.getQueryParameter("session_id").orEmpty()
        when (uri.getQueryParameter("status")) {
            "cancelled" -> {
                callback(FireblocksConnectResult.Cancelled)
                return
            }
            "error" -> {
                val code = uri.getQueryParameter("error_code") ?: "unknown"
                callback(FireblocksConnectResult.Error("connection failed", code, sessionId))
                return
            }
        }
        val address = uri.getQueryParameter("address").orEmpty()
        if (address.isEmpty()) {
            callback(FireblocksConnectResult.Error("malformed result"))
            return
        }
        callback(
            FireblocksConnectResult.Success(
                WalletConnection(
                    address = address,
                    chain = uri.getQueryParameter("chain").orEmpty(),
                    walletName = uri.getQueryParameter("walletName").orEmpty(),
                    walletImage = uri.getQueryParameter("walletImage").orEmpty(),
                    sessionId = sessionId,
                )
            )
        )
    }

    private fun buildUrl(flowURL: String, scheme: String, nonce: String): String {
        val sep = if (flowURL.contains("?")) "&" else "?"
        return flowURL + sep +
            "redirect_uri=" + Uri.encode("$scheme://$CALLBACK_HOST") +
            "&nonce=" + Uri.encode(nonce) +
            "&embedded=1"
    }

    /** Cryptographically-random nonce (CSRF correlation; not a secret). */
    private fun randomNonce(): String {
        val bytes = ByteArray(16)
        SecureRandom().nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

// ── Return catcher ─────────────────────────────────────────────────────────────

/**
 * Receives `<scheme>://wallet-callback`, hands it to [FireblocksConnect], and
 * bounces back to the app (dismissing the Custom Tab). Declare it in your
 * AndroidManifest with `launchMode="singleTask"` and an intent-filter for your
 * scheme + the `wallet-callback` host (see README).
 */
class FireblocksRedirectActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        intent?.data?.let { uri ->
            // Headless Phantom returns to <scheme>://phantom-headless — let the
            // headless engine consume it first; otherwise it's a visible-flow return.
            if (!FireblocksHeadlessConnect.handleReturnURL(uri)) {
                FireblocksConnect.handleRedirect(uri)
            }
        }
        // Bring the app's own UI back to the front, closing the Custom Tab.
        packageManager.getLaunchIntentForPackage(packageName)?.let {
            it.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
            startActivity(it)
        }
        finish()
    }
}

File: apps/connections/native/react-native/FireblocksConnect.ts
import * as WebBrowser from "expo-web-browser";
import * as Linking from "expo-linking";
import * as Crypto from "expo-crypto";

// ── Public API ──────────────────────────────────────────────────────────────

/** A wallet the user connected through the hosted Fireblocks flow. */
export interface WalletConnection {
  address: string;
  /** "evm" | "solana" */
  chain: string;
  walletName: string;
  /** Icon URL — often an SVG-sprite URL, so render it in a WebView `<img>`. */
  walletImage: string;
}

/** Thrown when the user dismisses the sheet. */
export class FireblocksConnectCancelled extends Error {}
/** Thrown on a nonce mismatch or an unparseable return. */
export class FireblocksConnectError extends Error {}

/**
 * Connect a self-custodial wallet through a hosted Fireblocks page — no SDK.
 *
 * ```ts
 * const wallet = await connectWallet({
 *   flowURL: "https://connect.example.com/",
 *   scheme: "myapp",              // must match your app.json "scheme"
 * });
 * console.log(wallet.address, wallet.chain);
 * ```
 *
 * Opens the page with `expo-web-browser`'s auth session
 * (`ASWebAuthenticationSession` on iOS, Chrome Custom Tabs on Android). It
 * appends `redirect_uri`, a random `nonce`, and `embedded=1`, verifies the
 * returned nonce, and resolves with the wallet.
 *
 * Most wallets return inside the session; some (Phantom) finish in their own
 * in-app browser and return via the scheme out-of-band — so we also listen on
 * `Linking` and dismiss the session when that happens. No extra wiring needed.
 */
export async function connectWallet(params: {
  flowURL: string;
  scheme: string;
}): Promise<WalletConnection> {
  const { flowURL, scheme } = params;
  const redirectUri = `${scheme}://wallet-callback`;
  const nonce = randomNonce();
  const url = withParams(flowURL, { redirect_uri: redirectUri, nonce, embedded: "1" });

  const callbackUrl = await openAndAwaitReturn(url, redirectUri);
  return verify(callbackUrl, nonce);
}

// ── Internals ─────────────────────────────────────────────────────────────

/** Race the auth session against a deep-link listener (out-of-band returns). */
function openAndAwaitReturn(url: string, redirectUri: string): Promise<string> {
  return new Promise((resolve, reject) => {
    let settled = false;
    const finish = (fn: () => void) => {
      if (settled) return;
      settled = true;
      subscription.remove();
      fn();
    };

    const subscription = Linking.addEventListener("url", ({ url: incoming }) => {
      if (incoming.startsWith(redirectUri)) {
        WebBrowser.dismissAuthSession(); // close the still-open sheet
        finish(() => resolve(incoming));
      }
    });

    // preferEphemeralSession skips the "<app> wants to use <domain> to sign in"
    // consent alert (no shared cookies), matching the native iOS behavior.
    WebBrowser.openAuthSessionAsync(url, redirectUri, { preferEphemeralSession: true })
      .then((result) => {
        if (result.type === "success") finish(() => resolve(result.url));
        else finish(() => reject(new FireblocksConnectCancelled("Cancelled")));
      })
      .catch((error) => finish(() => reject(error)));
  });
}

function verify(callbackUrl: string, expectedNonce: string): WalletConnection {
  const { queryParams } = Linking.parse(callbackUrl);
  const value = (key: string): string => {
    const v = queryParams?.[key];
    return typeof v === "string" ? v : "";
  };
  if (value("nonce") !== expectedNonce) {
    throw new FireblocksConnectError("Nonce mismatch"); // possible CSRF — reject
  }
  const address = value("address");
  if (!address) throw new FireblocksConnectError("Malformed result");
  return {
    address,
    chain: value("chain"),
    walletName: value("walletName"),
    walletImage: value("walletImage"),
  };
}

/** Append params without relying on RN's spotty URL/URLSearchParams support. */
function withParams(base: string, extra: Record<string, string>): string {
  const query = Object.entries(extra)
    .map(([k, v]) => `${k}=${encodeURIComponent(v)}`)
    .join("&");
  return base + (base.includes("?") ? "&" : "?") + query;
}

/** Cryptographically-random nonce (CSRF correlation). Not a secret, but must
 *  be unguessable. */
function randomNonce(): string {
  const bytes = Crypto.getRandomBytes(16);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}
